modules/solid_mechanics/src/materials/CombinedNonlinearHardeningPlasticity.C

```diff
@@ -0,0 +1,241 @@
+#include "CombinedNonlinearHardeningPlasticity.h"
+#include "Function.h"
+#include "ElasticityTensorTools.h"
+#include "RadialReturnBackstressStressUpdateBase.h"
+#include "ElasticityTensorTools.h"
+
+registerMooseObject("SolidMechanicsApp", ADCombinedNonlinearHardeningPlasticity);
+registerMooseObject("SolidMechanicsApp", CombinedNonlinearHardeningPlasticity);
+
+template <bool is_ad>
+InputParameters
+CombinedNonlinearHardeningPlasticityTempl<is_ad>::validParams()
+{
+  InputParameters params = RadialReturnBackstressStressUpdateBaseTempl<is_ad>::validParams();
+  params.addClassDescription(
+      "This class uses the discrete material in a radial return combined"
+      "plasticity model. This class can simulate both nonlinear isotropic and kinematic"
+      " hardening effects.");
+  params.addParam<Real>("_q", 0.0, "Saturation value for isotropic hardening (Q in Voce model)");
+  params.addParam<Real>("_b", 0.0, "Rate constant for isotropic hardening (b in Voce model)");
+  params.addParam<FunctionName>("yield_stress_function",
+                                "Yield stress as a function of temperature");
+  params.addParam<Real>("yield_stress", "The point at which plastic strain begins accumulating");
+  params.addParam<FunctionName>("isotropic_hardening_function",
+                                "True stress as a function of plastic strain");
+  params.addParam<Real>("isotropic_hardening_constant", "Isotropic hardening slope");
+  params.addCoupledVar("temperature", 0.0, "Coupled Temperature");
+  params.set<std::string>("effective_inelastic_strain_name") = "effective_plastic_strain";
+  params.addParam<Real>("kinematic_hardening_modulus", 0.0, "Kinematic hardening modulus");
+  params.addParam<Real>("kinematic_saturation_gamma",
```

Comment:
I told you to change this to say `kinematic_saturation` at the beginning of the name, but then it would be inconsistent with the naming scheme for `b` and `q`. Let's just keep it as `gamma`.